### PR TITLE
FISH-858 Upgrade to Hazelcast 4.2 from 4.1 and use Tenant Control APIs

### LINKED ISSUE #5159: No Current Invocation present if using Hazelcast backed cache from another module
# Description #
----------
Begining with Payara 5.2021.1 a quartz job started throwing exceptions during access to cache.

If all code from "job" module is moved to "core" module, the access to the cache from quartz job is OK.

## Expected Outcome ##
No exception

## Current Outcome ##

[2021-03-11T15:26:38.507+0100] [Payara 5.2021.1] [INFO] [] [] [tid: _ThreadID=770 _ThreadName=Scheduler_Worker-1] [timeMillis: 1615472798507] [levelValue: 800] [[
  15:26:38.506 [Scheduler_Worker-1] ERROR org.quartz.core.ErrorLogger - Job (DEFAULT.job-with-cache threw an exception.
org.quartz.SchedulerException: Job threw an unhandled exception.
	at org.quartz.core.JobRunShell.run(JobRunShell.java:234) [org.quartz-scheduler-quartz-1.8.6.jar:?]
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:549) [org.quartz-scheduler-quartz-1.8.6.jar:?]
Caused by: java.lang.IllegalStateException: No Current Invocation present
	at fish.payara.appserver.context.JavaEEContextUtilImpl.currentInvocation(JavaEEContextUtilImpl.java:90) ~[gf-web-connector.jar:?]
	at fish.payara.nucleus.hazelcast.contextproxy.CacheManagerProxy.getCache(CacheManagerProxy.java:87) ~[hazelcast-bootstrap.jar:?]
	at cz.i.quartz.JobUsingCache.getCache(JobUsingCache.java:43) ~[job_jar/:?]
	at cz.i.quartz.JobUsingCache.execute(JobUsingCache.java:60) ~[job_jar/:?]
	at org.quartz.core.JobRunShell.run(JobRunShell.java:223) ~[org.quartz-scheduler-quartz-1.8.6.jar:?]
	... 1 more
]]

## Steps to reproduce ## 

1. checkout https://github.com/sgflt/payara-test-case
2. deploy app
3. wait 10 seconds
4. exception will appear

## Environment ##

- **Distribution**: Full
- **JDK Version**:OpenJDK 11
- **Operating System**: Fedora 33

### PR DESCRIPTION
**Pull request by lprimak**
## Description
Context for this is hazelcast/hazelcast#10875 and all the issues mentioned within this issue as well.
Main feature is that Hazelcast Cache / Map / Queue etc. now work with Java EE constructs, and CDI beans, in Payara Data Grid as well.

fixes #5159 

### Blockers
Hz 4.2 needs to be released

### New tests
Got #759 / https://github.com/bgjug/jcache-workshop to work in clustered environment using Payara Micro and Server mix.
Most of this PR is removing a whole bunch of Proxy classes that are no longer needed

### Testing Performed
All current test suites pass

### Testing Environment
JDK 8, MacOS

## Documentation
There are new classes introduced for TenantControl, documented via javadoc in `payara-api`

### CODE REVIEW

**Comment by lprimak** [comment]
For reference, old PR https://github.com/payara/Payara/pull/4762

File: nucleus/payara-modules/hazelcast-bootstrap/src/main/java/fish/payara/nucleus/hazelcast/PayaraHazelcastTenant.java [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+                String componentId = contextInstance.getInstanceComponentId();
+                int unavailableCount = blockedCounts.compute(componentId, (k, v) -> v == null ? 0 : ++v);
+                log.log(unavailableCount > 100 ? Level.INFO : Level.FINEST,
+                        String.format("BLOCKED: tenant not available: %s, module %s, Operation: %s",
+                                componentId, moduleName, tenantable.getClass().getName()));
+                if (unavailableCount > 100) {
+                    blockedCounts.remove(componentId);
+                }
+                condition.await(100, TimeUnit.MILLISECONDS);
+            } catch (InterruptedException ex) {
```
Comment (MarkWareham) [inline nucleus/payara-modules/hazelcast-bootstrap/src/main/java/fish/payara/nucleus/hazelcast/PayaraHazelcastTenant.java:149]: are you intentionally swallowing the exception?

File: nucleus/payara-modules/hazelcast-bootstrap/src/main/java/fish/payara/nucleus/hazelcast/PayaraHazelcastTenant.java [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+            } finally {
+                lock.unlock();
+            }
+            return false;
+        }
+        return true;
```
Comment (MarkWareham) [inline nucleus/payara-modules/hazelcast-bootstrap/src/main/java/fish/payara/nucleus/hazelcast/PayaraHazelcastTenant.java:155]:
```suggestion
        if (!contextInstance.isLoaded()) {
            return true;
        }
        if (!tenantable.requiresTenantContext() || tenantNotRequired(tenantable)) {
            return true;
        }
        lock.lock();
        try {
            String componentId = contextInstance.getInstanceComponentId();
            int unavailableCount = blockedCounts.compute(componentId, (k, v) -> v == null ? 0 : ++v);
            log.log(unavailableCount > 100 ? Level.INFO : Level.FINEST,
                    String.format("BLOCKED: tenant not available: %s, module %s, Operation: %s",
                            componentId, moduleName, tenantable.getClass().getName()));
            if (unavailableCount > 100) {
                blockedCounts.remove(componentId);
            }
            condition.await(100, TimeUnit.MILLISECONDS);
        } catch (InterruptedException ex) {
        } finally {
            lock.unlock();
        }
        return false;        
```

**Review by MarkWareham — APPROVED** [review]
LGTM. Some suggestions, observations

**Comment by lprimak** [comment]
Waiting for Hazelcast 4.2 to be released in March

**Comment by lprimak** [comment]
jenkins test

**Comment by lprimak** [comment]
jenkins test

**Review by dmatej — CHANGES_REQUESTED** [review]
Not sure if it is alright to merge it to master, because I am sure it breaks something ... seems Metrics + Payara + Hazelcast have some common issue ... not sure if we can postpone that or if the fix should be part of this PR.
But I have a reproducer ... https://github.com/payara/Payara/pull/5019

```
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT: [2021-03-26T15:01:54.685+0000] [Payara 5.2021.2-SNAPSHOT] [WARNING] [] [javax.enterprise.web] [tid: _ThreadID=123 _ThreadName=http-thread-pool::http-listener-2(5)] [timeMillis: 1616770914685] [levelValue: 900] [[
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:   StandardWrapperValve[microprofile-metrics-resource]: Servlet.service() for servlet microprofile-metrics-resource threw exception
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT: java.lang.NullPointerException
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at fish.payara.appserver.context.JavaEEContextUtilImpl.isLoaded(JavaEEContextUtilImpl.java:149)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at fish.payara.appserver.context.JavaEEContextUtilImpl.isInvocationLoaded(JavaEEContextUtilImpl.java:118)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at fish.payara.nucleus.hazelcast.PayaraHazelcastTenantFactory.saveCurrentTenant(PayaraHazelcastTenantFactory.java:64)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at com.hazelcast.spi.impl.tenantcontrol.impl.TenantControlServiceImpl.initializeTenantControl(TenantControlServiceImpl.java:136)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at com.hazelcast.spi.impl.proxyservice.impl.ProxyRegistry.doCreateProxy(ProxyRegistry.java:234)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at com.hazelcast.spi.impl.proxyservice.impl.ProxyRegistry.createProxy(ProxyRegistry.java:219)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at com.hazelcast.spi.impl.proxyservice.impl.ProxyRegistry.getOrCreateProxyFuture(ProxyRegistry.java:186)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at com.hazelcast.spi.impl.proxyservice.impl.ProxyRegistry.getOrCreateProxy(ProxyRegistry.java:166)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at com.hazelcast.spi.impl.proxyservice.impl.ProxyServiceImpl.getDistributedObject(ProxyServiceImpl.java:153)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at com.hazelcast.instance.impl.HazelcastInstanceImpl.getDistributedObject(HazelcastInstanceImpl.java:354)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at com.hazelcast.instance.impl.HazelcastInstanceImpl.getMap(HazelcastInstanceImpl.java:173)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at com.hazelcast.instance.impl.HazelcastInstanceProxy.getMap(HazelcastInstanceProxy.java:95)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at fish.payara.nucleus.store.ClusteredStore.get(ClusteredStore.java:190)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at fish.payara.nucleus.microprofile.config.source.ClusterConfigSource.getValue(ClusterConfigSource.java:83)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at fish.payara.nucleus.microprofile.config.source.ClusterConfigSource.getOrdinal(ClusterConfigSource.java:74)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at fish.payara.nucleus.microprofile.config.spi.ConfigSourceComparator.compare(ConfigSourceComparator.java:54)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at fish.payara.nucleus.microprofile.config.spi.ConfigSourceComparator.compare(ConfigSourceComparator.java:50)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at java.util.TimSort.countRunAndMakeAscending(TimSort.java:355)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at java.util.TimSort.sort(TimSort.java:220)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at java.util.Arrays.sort(Arrays.java:1438)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at java.util.List.sort(List.java:478)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at java.util.Collections.sort(Collections.java:177)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at fish.payara.nucleus.microprofile.config.spi.PayaraConfig.<init>(PayaraConfig.java:106)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at fish.payara.nucleus.microprofile.config.spi.ConfigProviderResolverImpl.getConfig(ConfigProviderResolverImpl.java:280)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at fish.payara.nucleus.microprofile.config.spi.ConfigProviderResolverImpl.getConfig(ConfigProviderResolverImpl.java:306)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at fish.payara.nucleus.microprofile.config.spi.ConfigProviderResolverImpl.getConfig(ConfigProviderResolverImpl.java:213)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.eclipse.microprofile.config.ConfigProvider.getConfig(ConfigProvider.java:85)
16:01:54.688  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at fish.payara.microprofile.metrics.rest.MetricsResource.getGlobalTags(MetricsResource.java:175)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at fish.payara.microprofile.metrics.rest.MetricsResource.getOutputWriter(MetricsResource.java:162)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at fish.payara.microprofile.metrics.rest.MetricsResource.processRequest(MetricsResource.java:124)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at fish.payara.microprofile.metrics.rest.MetricsResource.doGet(MetricsResource.java:288)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at javax.servlet.http.HttpServlet.service(HttpServlet.java:645)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at javax.servlet.http.HttpServlet.service(HttpServlet.java:750)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.apache.catalina.core.StandardWrapper.service(StandardWrapper.java:1636)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:259)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:161)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:757)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:577)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at com.sun.enterprise.web.WebPipeline.invoke(WebPipeline.java:99)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:158)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.java:371)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:238)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:520)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:217)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:182)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:156)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:218)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:95)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:260)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:177)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:109)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:88)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:53)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:524)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:89)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:94)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:33)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:114)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:569)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:549)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT:        at java.lang.Thread.run(Thread.java:748)
16:01:54.689  INFO [docker-java-stream-176109586]                  D-PAYARA:64 - STDOUT: ]]
```

**Comment by dmatej** [comment]
Wow, too late :D

**Comment by MattGill98** [comment]
Does this need reverting @dmatej?

**Comment by dmatej** [comment]
I vote to continue with it if we fix it soon (next week). If not and it would break anything else, we will revert it.

**Comment by lprimak** [comment]
turned out a very simple fix. Good job @dmatej for discovering it.

**Pull request merged**